cmd/serverless/main.go

```diff
@@ -244,136 +249,68 @@ func runAgent(stopCh chan struct{}) (daemon *serverless.Daemon, err error) {
 
-	log.Debug("Enabling logs collection HTTP route")
-	if httpAddr, logsChan, err := daemon.EnableLogsCollection(); err != nil {
-		log.Error("While starting the HTTP Logs Server:", err)
-	} else {
-		// subscribe to the logs on the platform
-		if err := serverless.SubscribeLogs(serverlessID, httpAddr, logsType); err != nil {
-			log.Error("Can't subscribe to logs:", err)
+	waitingChan := make(chan bool, 2)
+
+	// starts logs collection
+	// ----------------------
+	go func(waitingChan chan bool) {
+		log.Debug("Enabling logs collection HTTP route")
+		logRegistrationURL := registration.BuildURL(os.Getenv(runtimeAPIEnvVar), logsAPIRegistrationRoute)
+		logRegistrationError := registration.EnableLogsCollection(
+			serverlessID,
+			logRegistrationURL,
+			logsAPIRegistrationTimeout,
+			os.Getenv(logsLogsTypeSubscribed),
+			logsAPIHttpServerPort,
+			logsAPICollectionRoute,
+			logsAPITimeout,
+			logsAPIMaxBytes,
+			logsAPIMaxItems)
+
+		if logRegistrationError != nil {
+			log.Error("Can't subscribe to logs:", logRegistrationError)
 		} else {
-			// we subscribed to the logs collection on the platform, let's instantiate
-			// a logs agent to collect/process/flush the logs.
-			if err := logs.StartServerless(
-				func() *autodiscovery.AutoConfig { return common.AC },
-				logsChan, nil,
-			); err != nil {
-				log.Error("Could not start an instance of the Logs Agent:", err)
-			}
+			setupLogAgent(logChannel)
 		}
-	}
-
-	// setup the forwarder, serializer and aggregator
-	// ----------------------------------------------
+		waitingChan <- true
+	}(waitingChan)
 
-	keysPerDomain, err := config.GetMultipleEndpoints()
-	if err != nil {
-		// we're not reporting the error to AWS because we don't want the function
-		// execution to be stopped. TODO(remy): discuss with AWS if there is way
-		// of reporting non-critical init errors.
-		log.Errorf("Misconfiguration of agent endpoints: %s", err)
-	}
-	forwarderTimeout := config.Datadog.GetDuration("forwarder_timeout") * time.Second
-	log.Debugf("Using a SyncForwarder with a %v timeout", forwarderTimeout)
-	f := forwarder.NewSyncForwarder(keysPerDomain, forwarderTimeout)
-	f.Start() //nolint:errcheck
-	serializer := serializer.NewSerializer(f, nil)
-
-	aggregatorInstance := aggregator.InitAggregator(serializer, nil, "serverless")
-	metricsChan := aggregatorInstance.GetBufferedMetricsWithTsChannel()
-
-	// prevents any UDP packets from being stuck in the buffer and not parsed during the current invocation
-	// by setting this option to 1ms, all packets received will directly be sent to the parser
-	config.Datadog.Set("dogstatsd_packet_buffer_flush_timeout", 1*time.Millisecond)
+	// starts trace agent
+	// ----------------------
+	traceAgent := &trace.ServerlessTraceAgent{}
+	go traceAgent.Start(config.Datadog.GetBool("apm_config.enabled"), &trace.LoadConfig{Path: datadogConfigPath}, waitingChan)
```

Comment:
I looked inside traceAgent.Start, and the code seems to be mostly synchronous, which means waitingChan is just telling us the function returned. That makes the class harder to initialize, and forces you to run go.traceAgent.Start in it's own goroutine, which isn't necessary. You could just replace it with a wait group eg:

```go
wg := sync.WaitGroup{}
wg.Add(1)
wg.Add(1)
go func() {
   defer wg.Done()
   traceAgent.Start(...)
}()
go func() {
   defer wg.Done()
   // ...
   registration.EnableLogsCollection(...)
}()
wg.Wait()
```